src/winconpty/winconpty.cpp

```diff
@@ -0,0 +1,368 @@
+    signalPacket[0] = PTY_SIGNAL_RESIZE_WINDOW;
+    signalPacket[1] = size.X;
+    signalPacket[2] = size.Y;
+
+    BOOL fSuccess = WriteFile(pPty->hSignal, signalPacket, sizeof(signalPacket), NULL, NULL);
+    return fSuccess ? S_OK : HRESULT_FROM_WIN32(GetLastError());
+}
+
+// Function Description:
+// - This closes each of the members of a PseudoConsole. It does not free the
+//      data associated with the PseudoConsole. This is helpful for testing,
+//      where we might stack allocate a PseudoConsole (instead of getting a
+//      HPCON via the API).
+// Arguments:
+// - pPty: A pointer to a PseudoConsole struct.
+// Return Value:
+// - <none>
+void _ClosePseudoConsoleMembers(_In_ PseudoConsole* pPty)
+{
+    if (pPty != NULL)
+    {
+        // See MSFT:19918626
+        // First break the signal pipe - this will trigger conhost to tear itself down
+        if (_HandleIsValid(pPty->hSignal))
+        {
+            CloseHandle(pPty->hSignal);
+            pPty->hSignal = 0;
+        }
+        // Then, wait on the conhost process before killing it.
+        // We do this to make sure the conhost finishes flushing any output it
+        //      has yet to send before we hard kill it.
+        if (_HandleIsValid(pPty->hConPtyProcess))
+        {
+            // If the conhost is already dead, then that's fine. Presumably
+            //      it's finished flushing it's output already.
+            DWORD dwExit = 0;
+            // If GetExitCodeProcess failed, it's likely conhost is already dead
+            //      If so, skip waiting regardless of whatever error
+            //      GetExitCodeProcess returned.
+            //      We'll just go straight to killing conhost.
+            if (GetExitCodeProcess(pPty->hConPtyProcess, &dwExit) && dwExit == STILL_ACTIVE)
+            {
+                WaitForSingleObject(pPty->hConPtyProcess, INFINITE);
+            }
+
+            TerminateProcess(pPty->hConPtyProcess, 0);
+            pPty->hConPtyProcess = 0;
+        }
+        // Then take care of the reference handle.
+        if (_HandleIsValid(pPty->hPtyReference))
+        {
+            CloseHandle(pPty->hPtyReference);
+            pPty->hPtyReference = 0;
+        }
+    }
+}
+
+// Function Description:
+// - This closes each of the members of a PseudoConsole, and HeapFree's the
+//      memory allocated to it. This should be used to cleanup any
+//      PseudoConosles that were created with CreatePseudoConsole.
+// Arguments:
+// - pPty: A pointer to a PseudoConsole struct.
+// Return Value:
+// - <none>
+VOID _ClosePseudoConsole(_In_ PseudoConsole* pPty)
+{
+    if (pPty != NULL)
+    {
+        _ClosePseudoConsoleMembers(pPty);
+        HeapFree(GetProcessHeap(), 0, pPty);
+    }
+}
+
+// These functions are defined in the console l1 apiset, which is generated from
+//      the consoleapi.apx file in minkernel\apiset\libs\Console.
+
+HRESULT CreatePseudoConsoleAsUser(_In_ HANDLE hToken,

```

Comment:
Did we ever make this public before? I guess we're not `__declspec(dllexport)`ing it from this dll, so probably nbd.

Reply:
It's not public in win32, but also not harmful. :smile: